File: core/agents/llm-abstraction/node/llm-client.ts
import { readFileSync, existsSync } from 'node:fs';
import { join } from 'node:path';

export interface LLMConfig {
  provider: string;
  model: string;
  apiKeyEnv: string;
}

export interface LLMProvider {
  complete(prompt: string, system?: string): Promise<string>;
}

export function loadConfig(configPath?: string): Record<string, any> {
  const path = configPath || join(process.cwd(), '.enterprise.yml');
  if (!existsSync(path)) {
    throw new Error(`.enterprise.yml not found at ${path}`);
  }
  // Simple YAML parsing for flat config
  const content = readFileSync(path, 'utf-8');
  const config: Record<string, any> = {};
  let section = '';
  let subsection = '';
  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    if (!line.startsWith(' ') && trimmed.endsWith(':')) {
      section = trimmed.slice(0, -1);
      config[section] = {};
      subsection = '';
    } else if (line.startsWith('  ') && !line.startsWith('    ') && trimmed.endsWith(':')) {
      subsection = trimmed.slice(0, -1);
      config[section][subsection] = {};
    } else if (trimmed.includes(':')) {
      const [key, ...rest] = trimmed.split(':');
      let value: any = rest.join(':').trim().replace(/^["']|["']$/g, '');
      if (value === 'true') value = true;
      else if (value === 'false') value = false;
      else if (/^\d+$/.test(value)) value = parseInt(value, 10);
      if (subsection && section) {
        config[section][subsection][key.trim()] = value;
      } else if (section) {
        config[section][key.trim()] = value;
      }
    }
  }
  return config;
}

export function getAgentConfig(role: string = 'default', configPath?: string): LLMConfig {
  const config = loadConfig(configPath);
  const agents = config.agents || {};
  const roleConfig = role !== 'default' && agents[role] ? agents[role] : agents.default || {};
  return {
    provider: roleConfig.provider || 'anthropic',
    model: roleConfig.model || 'claude-sonnet-4-20250514',
    apiKeyEnv: roleConfig.api_key_env || 'LLM_API_KEY',
  };
}

export async function getClient(role: string = 'default', configPath?: string): Promise<LLMProvider> {
  const config = getAgentConfig(role, configPath);
  const apiKey = config.apiKeyEnv ? process.env[config.apiKeyEnv] || '' : '';

  switch (config.provider) {
    case 'anthropic': {
      const { AnthropicProvider } = await import('./providers/anthropic.js');
      return new AnthropicProvider(config.model, apiKey);
    }
    case 'openai': {
      const { OpenAIProvider } = await import('./providers/openai.js');
      return new OpenAIProvider(config.model, apiKey);
    }
    case 'ollama': {
      const { OllamaProvider } = await import('./providers/ollama.js');
      return new OllamaProvider(config.model);
    }
    default:
      throw new Error(`Unknown LLM provider: ${config.provider}. Supported: anthropic, openai, ollama`);
  }
}

File: core/agents/llm-abstraction/python/llm_client.py
"""Provider-agnostic LLM client. Reads config from .enterprise.yml."""

import os
from pathlib import Path

try:
    import yaml
except ImportError:
    yaml = None


def _parse_yaml_simple(path: str) -> dict:
    """Minimal YAML parser for .enterprise.yml when PyYAML is not installed."""
    config = {}
    current_section = None
    current_subsection = None
    with open(path) as f:
        for line in f:
            stripped = line.strip()
            if not stripped or stripped.startswith("#"):
                continue
            if not line.startswith(" ") and stripped.endswith(":"):
                current_section = stripped[:-1]
                config[current_section] = {}
                current_subsection = None
            elif line.startswith("  ") and not line.startswith("    ") and stripped.endswith(":"):
                current_subsection = stripped[:-1]
                if current_section:
                    config[current_section][current_subsection] = {}
            elif ":" in stripped and current_section:
                key, _, value = stripped.partition(":")
                value = value.strip().strip('"').strip("'")
                if value.lower() == "true":
                    value = True
                elif value.lower() == "false":
                    value = False
                elif value.isdigit():
                    value = int(value)
                if current_subsection and current_section:
                    config[current_section].get(current_subsection, {})[key.strip()] = value
                elif current_section:
                    config[current_section][key.strip()] = value
    return config


def load_config(config_path: str = None) -> dict:
    """Load .enterprise.yml configuration."""
    if config_path is None:
        config_path = str(Path.cwd() / ".enterprise.yml")
    if not Path(config_path).exists():
        raise FileNotFoundError(f".enterprise.yml not found at {config_path}")
    if yaml:
        with open(config_path) as f:
            return yaml.safe_load(f)
    return _parse_yaml_simple(config_path)


def get_agent_config(role: str = "default", config_path: str = None) -> dict:
    """Get LLM config for a specific agent role."""
    config = load_config(config_path)
    agents = config.get("agents", {})
    if role != "default" and role in agents:
        return agents[role]
    return agents.get("default", {})


def get_client(role: str = "default", config_path: str = None):
    """Return an LLM client configured for the specified agent role."""
    agent_config = get_agent_config(role, config_path)
    provider = agent_config.get("provider", "anthropic")
    model = agent_config.get("model", "claude-sonnet-4-20250514")
    api_key_env = agent_config.get("api_key_env", "LLM_API_KEY")
    api_key = os.getenv(api_key_env, "") if api_key_env else None

    if provider == "anthropic":
        from .providers.anthropic_provider import AnthropicProvider
        return AnthropicProvider(model=model, api_key=api_key)
    elif provider == "openai":
        from .providers.openai_provider import OpenAIProvider
        return OpenAIProvider(model=model, api_key=api_key)
    elif provider == "ollama":
        from .providers.ollama_provider import OllamaProvider
        return OllamaProvider(model=model)
    else:
        raise ValueError(f"Unknown LLM provider: {provider}. Supported: anthropic, openai, ollama")
